File: src/main/java/emufog/reader/caida/CaidaFormatReader.java
/*
 * MIT License
 *
 * Copyright (c) 2018 emufog contributors
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
package emufog.reader.caida;

import emufog.config.Config;
import emufog.graph.AS;
import emufog.graph.Graph;
import emufog.graph.Node;
import emufog.reader.GraphReader;
import java.io.IOException;
import java.nio.charset.Charset;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * This reader can read in the CAIDA topology an build a graph based on that data.
 */
public class CaidaFormatReader implements GraphReader {

    private static final Logger LOG = LoggerFactory.getLogger(CaidaFormatReader.class);

    /**
     * initialize charset according to https://en.wikipedia.org/wiki/ISO/IEC_8859-1
     */
    private static final Charset CHARSET = StandardCharsets.ISO_8859_1;

    private static final int NODE_COLUMNS = 7;

    private static final int AS_COLUMNS = 3;

    private static final int EDGE_COLUMNS = 4;

    /* number of times AS field exceeds the Integer range */
    private int asNoInteger;

    /* number of times no node has been found with an ID to assign an AS */
    private int noNodeFoundForAS;

    /* number of times no node has been found with an ID to establish a connection */
    private int noNodeFoundForEdge;

    /* number of times ID field exceeds the Integer range */
    private int idsNoInteger;

    /* number of times a node line was skipped */
    private int nodeLineSkipped;

    /* number of times a AS line was skipped */
    private int asLineSkipped;

    /* number of times a link line was skipped */
    private int linkLineSkipped;

    /* number of times coordinate field exceeds the Float range */
    private int coordinatesNoFloats;

    /* mapping from ID's to coordinates of the nodes */
    private Map<Integer, Coordinates> nodeCoordinates;

    @Override
    public Graph readGraph(List<Path> files) throws IOException, IllegalArgumentException {
        if (files == null) {
            throw new IllegalArgumentException("No input files provided.");
        }
        Path nodesFile = getFileWithEnding(files, ".nodes.geo");
        if (nodesFile == null) {
            throw new IllegalArgumentException("The given files do not contain a .nodes.geo file.");
        }
        Path asFile = getFileWithEnding(files, ".nodes.as");
        if (asFile == null) {
            throw new IllegalArgumentException("The given files do not contain a .nodes.as file.");
        }
        Path linkFile = getFileWithEnding(files, ".links");
        if (linkFile == null) {
            throw new IllegalArgumentException("The given files do not contain a .links file.");
        }

        // initialize error counts
        asNoInteger = 0;
        coordinatesNoFloats = 0;
        idsNoInteger = 0;
        noNodeFoundForAS = 0;
        noNodeFoundForEdge = 0;
        nodeLineSkipped = 0;
        asLineSkipped = 0;
        linkLineSkipped = 0;

        nodeCoordinates = new HashMap<>();

        Graph graph = new Graph(Config.getConfig());

        // read in the nodes
        Files.lines(nodesFile, CHARSET).forEach(this::processNodeLine);

        // read in the AS
        Files.lines(asFile, CHARSET).forEach(x -> processASLine(graph, x));

        // read in the edges
        Files.lines(linkFile, CHARSET).forEach(x -> processLinkLine(graph, x));

        // log errors
        logResults();

        return graph;
    }

    /**
     * Logs the errors and not matching IDs while reading.
     */
    private void logResults() {
        // additional logging for debugging
        LOG.debug("ID out of Integer range: {}", idsNoInteger);
        LOG.debug("AS out of Integer range: {}", asNoInteger);
        LOG.debug("Coordinates out of Float range: {}", coordinatesNoFloats);
        LOG.debug("Number of times no nodes were found to assign an AS: {}", noNodeFoundForAS);
        LOG.debug("Number of times no nodes were found to build an edge: {}", noNodeFoundForEdge);
        LOG.debug("Nodes read without an AS: {}", nodeCoordinates.size());
        LOG.debug("Number of node lines skipped: {}", nodeLineSkipped);
        LOG.debug("Number of AS lines skipped: {}", asLineSkipped);
        LOG.debug("Number of link lines skipped: {}", linkLineSkipped);
    }

    /**
     * Reads in an edge of the graph.
     *
     * @param graph graph to add the edge to
     * @param line  current line to process
     */
    private void processLinkLine(Graph graph, String line) {
        String[] values = line.split(" ");
        if (values.length < EDGE_COLUMNS) {
            LOG.debug("There are not {} columns in the link line: {}", EDGE_COLUMNS, line);
            linkLineSkipped++;
            return;
        }

        String linkStr = values[1];
        linkStr = linkStr.substring(1, linkStr.length() - 1);
        int id;
        try {
            id = Integer.parseInt(linkStr);
        } catch (NumberFormatException e) {
            LOG.debug("Failed to parse the link id {} to an integer.", linkStr);
            idsNoInteger++;
            return;
        }

        for (int i = 3; i < values.length - 1; ++i) {
            String sourceStr = values[i];
            int end = sourceStr.indexOf(':');
            if (end == -1) {
                end = sourceStr.length();
            }
            sourceStr = sourceStr.substring(1, end);
            int sourceID;
            try {
                sourceID = Integer.parseInt(sourceStr);
            } catch (NumberFormatException e) {
                LOG.debug("Failed to parse the link's source id {} to an integer.", sourceStr);
                idsNoInteger++;
                return;
            }

            String destinationStr = values[i + 1];
            end = destinationStr.indexOf(':');
            if (end == -1) {
                end = destinationStr.length();
            }
            destinationStr = destinationStr.substring(1, end);
            int destinationID;
            try {
                destinationID = Integer.parseInt(destinationStr);
            } catch (NumberFormatException e) {
                LOG.debug("Failed to parse the link's destination id {} to an integer.", destinationStr);
                idsNoInteger++;
                return;
            }

            Node from = graph.getEdgeNode(sourceID);
            Node to = graph.getEdgeNode(destinationID);

            if (from != null && to != null) {
                graph.createEdge(id, from, to, getLatency(from, to), 1000);
            } else {
                LOG.debug("To create a link source and destination must be found.");
                noNodeFoundForEdge++;
            }
        }
    }

    /**
     * Adapts the AS field of the node identified in the current line.
     *
     * @param graph graph to modify the nodes from
     * @param line  current line to process
     */
    private void processASLine(Graph graph, String line) {
        String[] values = line.split(" ");
        if (values.length < AS_COLUMNS) {
            asLineSkipped++;
            LOG.debug("There are not {} columns in the autonomous system line: {}", AS_COLUMNS, line);
            return;
        }

        String nodeStr = values[1].substring(1);
        int id;
        try {
            id = Integer.parseInt(nodeStr);
        } catch (NumberFormatException e) {
            LOG.debug("Failed to parse the id {} to an integer.", nodeStr);
            idsNoInteger++;
            return;
        }

        int as;
        try {
            as = Integer.parseInt(values[2]);
        } catch (NumberFormatException e) {
            LOG.debug("Failed to parse the autonomous system id {} to an integer.", values[2]);
            asNoInteger++;
            return;
        }

        AS system = graph.getOrCreateAutonomousSystem(as);
        graph.createEdgeNode(id, system);
    }

    /**
     * Reads in and process a line of the input file to add a node to the graph given.
     *
     * @param line current line to process
     */
    private void processNodeLine(String line) {
        String[] values = line.split("\t");
        if (values.length < NODE_COLUMNS) {
            LOG.debug("There are not {} columns in the node line: {}", NODE_COLUMNS, line);
            nodeLineSkipped++;
            return;
        }

        String nodeStr = values[0];
        nodeStr = nodeStr.substring(10, nodeStr.length() - 1);
        int id;
        try {
            id = Integer.parseInt(nodeStr);
        } catch (NumberFormatException e) {
            LOG.debug("Failed to parse the id {} to an integer.", nodeStr);
            idsNoInteger++;
            return;
        }

        try {
            float xPos = Float.parseFloat(values[5]);
            float yPos = Float.parseFloat(values[6]);
            nodeCoordinates.put(id, new Coordinates(xPos, yPos));
        } catch (NumberFormatException e) {
            LOG.debug("Failed to parse coordinates {} and {} to floats.", values[5], values[6]);
            coordinatesNoFloats++;
        }
    }

    /**
     * Returns the path of the file matching the giving file extension at the end.
     *
     * @param files         list of possible files
     * @param fileExtension file extension to match
     * @return the file of the list matching the extension or {@code null} if not found
     */
    private static Path getFileWithEnding(List<Path> files, String fileExtension) {
        return files.stream()
            .filter(f -> f != null && f.toString().endsWith(fileExtension))
            .findFirst()
            .orElse(null);
    }

    private float getLatency(Node from, Node to) {
        return 1.f;
    }

    /**
     * Coordinates of a node in the graph. This class gets mapped to the respective ID of the node.
     */
    static class Coordinates {

        /**
         * x coordinate
         */
        final float xPos;

        /**
         * y coordinate
         */
        final float yPos;

        /**
         * Creates a new coordinate object in a 2D coordinate system.
         *
         * @param x x coordinate
         * @param y y coordinate
         */
        Coordinates(float x, float y) {
            xPos = x;
            yPos = y;
        }
    }
}

File: src/main/kotlin/emufog/reader/caida/CaidaFormatReader.kt
/*
 * MIT License
 *
 * Copyright (c) 2018 emufog contributors
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
package emufog.reader.caida

import emufog.config.Config
import emufog.graph.EdgeNode
import emufog.graph.Graph
import emufog.graph.Node
import emufog.reader.GraphReader
import emufog.util.IDManager
import org.slf4j.LoggerFactory
import java.nio.charset.StandardCharsets
import java.nio.file.Path

/**
 * This reader can read in the [CAIDA](https://www.caida.org/) topology an build a graph based on that data.
 */
object CaidaFormatReader : GraphReader {

    /**
     * Reads in a graph from a Caida dataset. Such a dataset consists of a `.nodes.geo`, a `.nodes.as` and a `.links`
     * file. The list of input files need to contain those three files.
     *
     * @param files the list of files to read in
     * @return read in graph object
     */
    override fun readGraph(files: List<Path>): Graph = CaidaFormatReaderImpl(files).readGraph()
}

private class CaidaFormatReaderImpl(files: List<Path>) {

    private companion object {

        private val LOG = LoggerFactory.getLogger(CaidaFormatReader::class.java)

        /**
         * initialize charset according to https://en.wikipedia.org/wiki/ISO/IEC_8859-1
         */
        private val CHARSET = StandardCharsets.ISO_8859_1

        private const val NODE_COLUMNS = 8

        private const val AS_COLUMNS = 3

        private const val EDGE_COLUMNS = 4
    }

    /* number of times AS field exceeds the Integer range */
    private var asNoInteger = 0

    /* number of times no node has been found with an ID to assign an AS */
    private var noNodeFoundForAS = 0

    /* number of times no node has been found with an ID to establish a connection */
    private var noNodeFoundForEdge = 0

    /* number of times ID field exceeds the Integer range */
    private var idsNoInteger = 0

    /* number of times a node line was skipped */
    private var nodeLineSkipped = 0

    /* number of times a AS line was skipped */
    private var asLineSkipped = 0

    /* number of times a link line was skipped */
    private var linkLineSkipped = 0

    /* number of times coordinate field exceeds the Float range */
    private var coordinatesNoFloats = 0

    /* mapping from ID's to coordinates of the nodes */
    private val nodeCoordinates: MutableMap<Int, Coordinates> = HashMap()

    private val nodesFile = getFileWithEnding(files, ".nodes.geo")
        ?: throw IllegalArgumentException("The given files do not contain a .nodes.geo file.")

    private val asFile = getFileWithEnding(files, ".nodes.as")
        ?: throw IllegalArgumentException("The given files do not contain a .nodes.as file.")

    private val linkFile = getFileWithEnding(files, ".links")
        ?: throw IllegalArgumentException("The given files do not contain a .links file.")

    private val spaceRegex = "\\s".toRegex()

    private val idManager = IDManager()

    private val graph = Graph(Config.config!!)

    /**
     * Reads in and returns a [Graph] object from the first `.nodes.geo`, `.nodes.as` and `.links` file. Skips all the
     * lines that can not be processed. Such cases are counted and printed to the debug output after reading in the
     * graph.
     */
    internal fun readGraph(): Graph {
        // read in the nodes
        nodesFile.toFile().forEachLine(CHARSET) { processNodeLine(it) }

        // read in the AS
        asFile.toFile().forEachLine(CHARSET) { processASLine(it) }

        // read in the edges
        linkFile.toFile().forEachLine(CHARSET) { processLinkLine(it) }

        // log errors
        logResults()

        return graph
    }

    /**
     * Logs the errors and not matching IDs while reading.
     */
    private fun logResults() {
        // additional logging for debugging
        LOG.debug("ID out of Integer range: {}", idsNoInteger)
        LOG.debug("AS out of Integer range: {}", asNoInteger)
        LOG.debug("Coordinates out of Float range: {}", coordinatesNoFloats)
        LOG.debug("Number of times no nodes were found to assign an AS: {}", noNodeFoundForAS)
        LOG.debug("Number of times no nodes were found to build an edge: {}", noNodeFoundForEdge)
        LOG.debug("Nodes read without an AS: {}", nodeCoordinates.size)
        LOG.debug("Number of node lines skipped: {}", nodeLineSkipped)
        LOG.debug("Number of AS lines skipped: {}", asLineSkipped)
        LOG.debug("Number of link lines skipped: {}", linkLineSkipped)
    }

    /**
     * Reads in an edge of the graph.
     */
    private fun processLinkLine(line: String) {
        if (line.startsWith('#')) {
            return
        }
        val values = line.split(spaceRegex)
        if (values.size < EDGE_COLUMNS) {
            LOG.debug("There are not {} columns in the link line: {}", EDGE_COLUMNS, line)
            linkLineSkipped++
            return
        }

        val linkStr = extractIdFromString(values[1])
        val id = linkStr.toIntOrNull()
        if (id == null) {
            LOG.debug("Failed to parse the link id {} to an integer.", linkStr)
            idsNoInteger++
            return
        }

        val sourceStr = extractIdFromString(values[3])
        val sourceID = sourceStr.toIntOrNull()
        if (sourceID == null) {
            LOG.debug("Failed to parse the link's source id {} to an integer.", sourceStr)
            idsNoInteger++
            return
        }

        for (i in 4 until values.size) {
            val destinationStr = extractIdFromString(values[i])
            val destinationID = destinationStr.toIntOrNull()
            if (destinationID == null) {
                LOG.debug("Failed to parse the link's destination id {} to an integer.", destinationStr)
                idsNoInteger++
                return
            }

            val from: EdgeNode? = graph.getEdgeNode(sourceID)
            val to: EdgeNode? = graph.getEdgeNode(destinationID)

            if (from == null || to == null) {
                LOG.debug("To create a link source and destination must be found.")
                noNodeFoundForEdge++
                return
            }

            var checkedId: Int = id
            if (idManager.isUsed(id)) {
                checkedId = idManager.getNextID()
                LOG.debug("The original edge ID: {} is already in use. Using {} instead.", id, checkedId)
            }
            graph.createEdge(checkedId, from, to, getLatency(from, to), 1000f)
            idManager.setUsed(checkedId)
        }
    }

    /**
     * Adapts the AS field of the node identified in the current line.
     */
    private fun processASLine(line: String) {
        if (line.startsWith('#')) {
            return
        }
        val values = line.split(spaceRegex)
        if (values.size < AS_COLUMNS) {
            asLineSkipped++
            LOG.debug("There are not {} columns in the autonomous system line: {}", AS_COLUMNS, line)
            return
        }

        val nodeStr = extractIdFromString(values[1])
        val id = nodeStr.toIntOrNull()
        if (id == null) {
            LOG.debug("Failed to parse the id {} to an integer.", nodeStr)
            idsNoInteger++
            return
        }

        val system = values[2].toIntOrNull()
        if (system == null) {
            LOG.debug("Failed to parse the autonomous system id {} to an integer.", values[2])
            asNoInteger++
            return
        }

        graph.createEdgeNode(id, graph.getOrCreateAutonomousSystem(system))
    }

    /**
     * Reads in and process a line of the input file to add a node to the graph given.
     */
    private fun processNodeLine(line: String) {
        if (line.startsWith('#')) {
            return
        }
        val values = line.split(spaceRegex)
        if (values.size < NODE_COLUMNS) {
            LOG.debug("There are not {} columns in the node line: {}", NODE_COLUMNS, line)
            nodeLineSkipped++
            return
        }

        val nodeStr = extractIdFromString(values[1])
        val id = nodeStr.toIntOrNull()
        if (id == null) {
            LOG.debug("Failed to parse the id {} to an integer.", nodeStr)
            idsNoInteger++
            return
        }

        try {
            nodeCoordinates[id] = Coordinates(values[6].toFloat(), values[7].toFloat())
        } catch (e: NumberFormatException) {
            LOG.debug("Failed to parse coordinates {} and {} to floats.", values[6], values[7])
            coordinatesNoFloats++
        }
    }

    private fun getLatency(from: Node, to: Node): Float {
        return 1F
    }

    private fun extractIdFromString(s: String): String {
        var end = s.indexOfFirst { it == ':' }
        if (end == -1) {
            end = s.length
        }

        return s.substring(1, end)
    }
}

/**
 * Coordinates of a node in the graph. This class gets mapped to the respective ID of the node.
 *
 * @property xPos x coordinate
 * @property yPos y coordinate
 */
private data class Coordinates(val xPos: Float, val yPos: Float)

/**
 * Returns the path of the file matching the giving file extension at the end.
 *
 * @param files list of possible files
 * @param fileExtension file extension to match
 * @return the file of the list matching the extension or `null` if not found
 */
private fun getFileWithEnding(files: List<Path>, fileExtension: String): Path? {
    return files.firstOrNull { it.toString().endsWith(fileExtension) }
}
